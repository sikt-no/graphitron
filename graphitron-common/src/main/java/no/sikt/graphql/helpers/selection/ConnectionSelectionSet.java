package no.sikt.graphql.helpers.selection;

import graphql.schema.DataFetchingFieldSelectionSet;

import java.util.List;
import java.util.regex.Pattern;

/**
 * A helper class for generated code. Helps with checking the selection set from the {@link graphql.schema.DataFetchingEnvironment}
 * for fields used in queries that use the connection types that follow the
 * <a href="https://relay.dev/graphql/connections.htm">cursor connection specification</a>.
 */
public class ConnectionSelectionSet extends SelectionSet {
    private final static String CONNECTION_NODE_REGEX_PATTERN_PREFIX = "^(edges/node/|nodes/|)";

    public ConnectionSelectionSet(List<DataFetchingFieldSelectionSet> selectionSets) {
        super(selectionSets);
    }

    public ConnectionSelectionSet(DataFetchingFieldSelectionSet selectionSet) {
        super(selectionSet);
    }

    public boolean contains(String path) {
        var pattern = Pattern.compile(CONNECTION_NODE_REGEX_PATTERN_PREFIX + path);
        return selectionSet.stream().anyMatch(it -> pattern.matcher(it).matches());
    }
}
