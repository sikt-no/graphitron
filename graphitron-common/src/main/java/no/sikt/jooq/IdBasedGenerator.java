package no.sikt.jooq;


import org.jooq.codegen.GeneratorStrategy;
import org.jooq.codegen.JavaGenerator;
import org.jooq.codegen.JavaWriter;
import org.jooq.meta.ColumnDefinition;
import org.jooq.meta.TableDefinition;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

//TODO Denne klassen skal endres (kanskje fjernes) i forbindelse med https://unit.atlassian.net/browse/FSP-266
public class IdBasedGenerator extends JavaGenerator {

    private static final Pattern toUpperPattern = Pattern.compile("_\\w|^\\w");

    @Override
    public void generateTableClassFooter(TableDefinition table, JavaWriter out) {
        super.generateTableClassFooter(table, out);

        if (table.getPrimaryKey() == null) {
            return;
        }

        List<ColumnDefinition> keyColumns = table.getPrimaryKey()
                .getKeyColumns();
        var primaryKeyFields = resolveFieldNames(keyColumns);

        generateGetter(out, primaryKeyFields, "Id", "Id");
        generateFieldsGetter(out, table, primaryKeyFields);

        var qualifiers = new ArrayList<Map.Entry<String, String>>();

        for (var fk : table.getForeignKeys()) {
            var name = fk.getKeyColumns().get(0).getName();
            var qualifier = toUpperPattern.matcher(name).replaceAll(snakeToCamel());
            qualifiers.add(Map.entry(fk.getName(), qualifier));

            var sourceFieldNames = resolveFieldNames(fk.getKeyColumns());
            generateGetter(out, sourceFieldNames, qualifier, name);
            generateHasCondition(out, sourceFieldNames, qualifier);
        }
        generateQualifierMap(out, qualifiers);
    }

    @Override
    protected void generateRecordClassFooter(TableDefinition table, JavaWriter out) {

        if (table.getPrimaryKey() == null) {
            return;
        }
        super.generateRecordClassFooter(table, out);
        generateRecordClassIdGetter(out, table.getPrimaryKey().getKeyColumns().size());
        generateRecordClassIdSetter(out);
    }

    private void generateQualifierMap(JavaWriter out, List<Map.Entry<String, String>> qualifiers) {
        out.println();
        if (qualifiers.isEmpty()) {
            out.println("private static final java.util.Map<String, String> qualifiers = java.util.Map.of();");
        } else {
            out.println("private static final java.util.Map<String, String> qualifiers = java.util.Map.ofEntries(");
            for (int i = 0; i < qualifiers.size() - 1; i++) {
                var entry = qualifiers.get(i);
                out.println(String.format("java.util.Map.entry(\"%s\", \"%s\"),", entry.getKey(), entry.getValue()));
            }

            var lastEntry = qualifiers.get(qualifiers.size() - 1);
            out.println(String.format("java.util.Map.entry(\"%s\", \"%s\")", lastEntry.getKey(), lastEntry.getValue()));
            out.println(");");
        }

        out.println();
        out.println("public String getQualifier(String keyName) { return qualifiers.get(keyName); }");
    }

    private List<String> resolveFieldNames(List<ColumnDefinition> keyColumns) {
        return keyColumns.stream()
                .map(column -> getStrategy().getJavaIdentifier(column))
                .collect(Collectors.toList());
    }

    private void generateGetter(JavaWriter out, List<String> primaryKeyFieldNames, String qualifier, String unqualifiedName) {
        var args = IntStream.range(0, primaryKeyFieldNames.size())
                .mapToObj(i -> "s" + i)
                .collect(Collectors.joining(", "));

        var toStrs = IntStream.range(0, primaryKeyFieldNames.size())
                .mapToObj(i -> "s" + i + ".toString()")
                .collect(Collectors.joining("+ \",\" +"));

        out.println();
        out.println("public org.jooq.SelectField<String> get%s() {", qualifier);
        out.println("return DSL.row(%s)", String.join(", ", primaryKeyFieldNames));
        out.tab(1).println(".mapping(String.class, org.jooq.Functions.nullOnAnyNull((%s) -> %s))",
                args, toStrs);
        out.tab(1).println(".as(DSL.name(\"%s\"));", unqualifiedName);
        out.println("}");
    }

    private void generateFieldsGetter(JavaWriter out, TableDefinition table, List<String> fields) {
        var recordType = getStrategy().getFullJavaClassName(table, GeneratorStrategy.Mode.RECORD);

        out.println();
        out.println("public java.util.List<TableField<%s, ?>> getIdFields() {", recordType);
        out.println("return java.util.List.of(%s);", String.join(", ", fields));
        out.println("}");
    }

    private void generateHasCondition(JavaWriter out, List<String> fieldNames, String qualifier) {
        out.println();
        out.println("public org.jooq.Condition has%s(String id) {", qualifier);
        out.println("return has%ss(java.util.Set.of(id));", qualifier);
        out.println("}");
        out.println();
        out.println("public org.jooq.Condition has%ss(java.util.Set<String> ids) {", qualifier);
        out.println("var field = java.util.List.of(%s).get(0);", String.join(", ", fieldNames));
        out.println("var converted = ids.stream().map(it -> field.getDataType().convert(it)).collect(java.util.stream.Collectors.toList());");

        out.println("return field.in(converted);");
        out.println("}");
    }

    private Function<MatchResult, String> snakeToCamel() {
        return match -> match.group().toUpperCase().replaceAll("_", "");
    }

    private void generateRecordClassIdGetter(JavaWriter out, int noOfPrimaryKeyFields) {
        out.println();
        out.println("public String getId() {");
        out.print("return \"\" + ");

        for (int i = 0; i < noOfPrimaryKeyFields; i++) {
            if (i > 0) out.print(" + \",\" + ");
            out.print("(Long) get(%s)", i);
        }
        out.println(";");
        out.println("}");
    }

    private void generateRecordClassIdSetter(JavaWriter out) {
        out.println();
        out.println("public void setId(String id) {");
        out.println("String[] split = id.split(\",\");");
        out.println("for (int i = 0; i < split.length; i++) {");
        out.println("set(i, split[i]);");
        out.println("}");
        out.println("}");
    }
}
