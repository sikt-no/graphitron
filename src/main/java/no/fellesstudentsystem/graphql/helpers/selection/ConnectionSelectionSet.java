package no.fellesstudentsystem.graphql.helpers.selection;

import graphql.schema.DataFetchingFieldSelectionSet;

import java.util.List;
import java.util.regex.Pattern;

public class ConnectionSelectionSet extends SelectionSet {
    private final static String CONNECTION_NODE_REGEX_PATTERN_PREFIX = "^(edges/node/|nodes/|)";

    public ConnectionSelectionSet(List<DataFetchingFieldSelectionSet> selectionSets) {
        super(selectionSets);
    }

    public ConnectionSelectionSet(DataFetchingFieldSelectionSet selectionSet) {
        super(selectionSet);
    }

    public boolean contains(String path) {
        var pattern = Pattern.compile(CONNECTION_NODE_REGEX_PATTERN_PREFIX + path);
        return selectionSet.stream().anyMatch(it -> pattern.matcher(it).matches());
    }
}
