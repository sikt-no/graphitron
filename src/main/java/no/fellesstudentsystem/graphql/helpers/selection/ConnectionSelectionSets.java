package no.fellesstudentsystem.graphql.helpers.selection;

import graphql.schema.DataFetchingFieldSelectionSet;

import java.util.List;

public class ConnectionSelectionSets extends SelectionSets {
    private final static String CONNECTION_NODE_GLOB_PATTERN_PREFIX = "{edges/node/,nodes/,}";

    public ConnectionSelectionSets(List<DataFetchingFieldSelectionSet> selectionSets) {
        super(selectionSets);
    }

    public ConnectionSelectionSets(DataFetchingFieldSelectionSet selectionSet) {
        super(selectionSet);
    }

    public boolean contains(String path) {
        return selectionSets.stream().anyMatch(it -> it.contains(CONNECTION_NODE_GLOB_PATTERN_PREFIX + path));
    }
}
