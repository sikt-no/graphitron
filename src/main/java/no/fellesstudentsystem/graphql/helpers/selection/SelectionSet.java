package no.fellesstudentsystem.graphql.helpers.selection;

import graphql.schema.DataFetchingFieldSelectionSet;
import graphql.schema.SelectedField;
import org.jooq.Field;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import static org.jooq.impl.DSL.inline;
import static org.jooq.impl.DSL.noField;

/**
 * A helper class for generated code. Helps with checking the selection set from the {@link graphql.schema.DataFetchingEnvironment}
 * for fields used in queries and mutations.
 */
public class SelectionSet {
    protected final Set<String> selectionSet;
    private final String prefix;

    public SelectionSet(List<DataFetchingFieldSelectionSet> selectionSets) {
        this.selectionSet = selectionSets.stream()
                .flatMap(selectionSet -> selectionSet.getFields().stream())
                .map(SelectedField::getQualifiedName)
                .collect(Collectors.toSet());
        prefix = "";
    }

    public SelectionSet(DataFetchingFieldSelectionSet selectionSet) {
        this(List.of(selectionSet));
    }

    private SelectionSet(Set<String> selectionSet, String prefix) {
        this.selectionSet = selectionSet;
        this.prefix = prefix + "/";
    }

    public SelectionSet withPrefix(String prefix) {
        return new SelectionSet(selectionSet, prefix);
    }

    public <T> Field<T> optional(String path, Field<T> field, T _default) {
        return contains(path) ? field : noField(inline(_default));
    }

    public <T> Field<T> optional(String path, Field<T> field) {
        return contains(path) ? field : noField(inline((T) null));
    }

    public boolean contains(String path) {
        return selectionSet.stream().anyMatch(it -> it.contains(prefix + path));
    }
}