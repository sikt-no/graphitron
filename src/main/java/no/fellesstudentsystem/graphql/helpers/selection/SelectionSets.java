package no.fellesstudentsystem.graphql.helpers.selection;

import graphql.schema.DataFetchingFieldSelectionSet;
import org.jooq.Field;

import java.util.List;

import static org.jooq.impl.DSL.inline;
import static org.jooq.impl.DSL.noField;

public class SelectionSets {
    protected final List<DataFetchingFieldSelectionSet> selectionSets;
    private final String prefix;

    public SelectionSets(List<DataFetchingFieldSelectionSet> selectionSets) {
        this.selectionSets = selectionSets;
        prefix = "";
    }

    public SelectionSets(DataFetchingFieldSelectionSet selectionSet) {
        this.selectionSets = List.of(selectionSet);
        prefix = "";
    }

    private SelectionSets(List<DataFetchingFieldSelectionSet> selectionSets, String prefix) {
        this.selectionSets = selectionSets;
        this.prefix = prefix + "/";
    }

    public SelectionSets withPrefix(String prefix) {
        return new SelectionSets(selectionSets, prefix);
    }

    public <T> Field<T> optional(String path, Field<T> field, T _default) {
        return contains(path) ? field : noField(inline(_default));
    }

    public <T> Field<T> optional(String path, Field<T> field) {
        return contains(path) ? field : noField(inline((T) null));
    }

    public boolean contains(String path) {
        return selectionSets.stream().anyMatch(it -> it.contains(prefix + path));
    }
}