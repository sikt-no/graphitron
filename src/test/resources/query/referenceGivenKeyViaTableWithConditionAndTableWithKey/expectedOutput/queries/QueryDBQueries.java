package fake.code.generated.queries.query;

import static no.sikt.graphitron.jooq.generated.testdata.Keys.*;
import static no.sikt.graphitron.jooq.generated.testdata.Tables.*;

import fake.graphql.example.package.model.Store;
import java.lang.String;
import java.util.List;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import org.jooq.DSLContext;
import org.jooq.Functions;
import org.jooq.impl.DSL;

public class QueryDBQueries {
    public List<Store> storeForQuery(DSLContext ctx, String id, SelectionSet select) {
        return ctx
                .select(
                        DSL.row(
                                STORE.getId().as("id")
                        ).mapping(Functions.nullOnAllNull(Store::new)).as("store")
                )
                .from(STORE)
                .where(STORE.ID.eq(id))
                .orderBy(STORE.getIdFields())
                .fetch(0, Store.class);
    }
}