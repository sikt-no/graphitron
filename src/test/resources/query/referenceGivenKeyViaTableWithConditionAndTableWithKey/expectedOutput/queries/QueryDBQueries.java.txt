package fake.code.generated.queries.query;

import static no.sikt.graphitron.jooq.generated.testdata.FieldHelpers.*;
import static no.sikt.graphitron.jooq.generated.testdata.Keys.*;
import static no.sikt.graphitron.jooq.generated.testdata.Tables.*;
import static org.jooq.Functions.*;
import static org.jooq.impl.DSL.*;

import fake.graphql.example.package.model.Store;
import java.lang.String;
import java.util.List;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import org.jooq.DSLContext;

public class QueryDBQueries {
    public List<Store> storeForQuery(DSLContext ctx, String id, SelectionSet select) {
        return ctx
                .select(
                        row(
                                STORE.getId().as("id")
                        ).mapping(nullOnAllNull(Store::new)).as("store")
                )
                .from(STORE)
                .where(STORE.ID.eq(id))
                .orderBy(STORE.getIdFields())
                .fetch(0, Store.class);
    }
}